solr/distribution/build.gradle

```diff
@@ -0,0 +1,141 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.apache.lucene.gradle.Checksum
+
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+
+plugins {
+  id 'signing'
+}
+
+// This project puts together the Solr "distribution", assembling bits and pieces
+// from across the project structure into release artifacts.
+
+ext {
+  releaseDir = file("${buildDir}/release")
+  withSignedArtifacts = { ->
+    def propValue = propertyOrDefault("sign", null)
+    // Allow -Psign to work as a shorthand for -Psign=true
+    return propValue != null && (propValue.isBlank() || Boolean.parseBoolean(propValue))
+  }.call()
+}
+
+// Prepare the "source" distribution artifact.
+apply from: buildscript.sourceFile.toPath().resolveSibling("source-release.gradle")
+
+// Prepare the "binary" distribution artifact.
+apply from: buildscript.sourceFile.toPath().resolveSibling("binary-release.gradle")
+
+// Configure artifact signing.
+apply from: buildscript.sourceFile.toPath().resolveSibling("artifact-signing.gradle")
+
+
+// Set up the HTML-rendered "changes" distribution artifact by linking to documentation's output.
+configurations {
+  changesHtml
+  solrTgz
+  solrTgzSignature
+  docker
+}
+
+artifacts {
+  solrTgz assembleBinaryTgz
+  solrTgzSignature file(assembleBinaryTgz.archiveFileName.get() + ".asc")
+}
+
+dependencies {
+  changesHtml project(path: ":solr:documentation", configuration: "changesHtml")
+  docker project(path: ':solr:docker', configuration: 'packaging')
+}
+
+// Compute checksums for release archives.
+task computeChecksums(type: Checksum) {
+  algorithm = Checksum.Algorithm.SHA512
+
+  files = objects.fileCollection()
+  [
+      tasks.assembleSourceTgz,
+      tasks.assembleBinaryTgz,
+      tasks.assembleBinaryZip,
+  ].each { dep ->
+    dependsOn dep
+    files += dep.outputs.files
+  }
+
+  outputDir = file("${buildDir}/checksums")
+}
+
+
+task prepareGitRev() {
+  dependsOn ":gitStatus"
+
+  ext.outputFile = file("${buildDir}/.gitrev")
+
+  outputs.file(ext.outputFile)
+  inputs.property("gitrev", provider { -> rootProject.ext.gitRev })
+
+  doFirst {
+    Files.writeString(ext.outputFile.toPath(), rootProject.ext.gitRev, StandardCharsets.UTF_8)
+  }
+}
+
+
+// Assemble everything needed in the release folder structure.
+task assembleRelease(type: Sync) {
+  description "Assemble all Solr artifacts for a release."
+  dependsOn rootProject.getTasksByName("mavenToLocalFolder", true)[0]
+
+  from(configurations.changesHtml, {
+    into "changes"
+  })
+
+  from(configurations.docker, {
+    into "docker"
```

Comment:
I added in a `docker` folder but I don't think we need it in the release process because
* The same content is inside the release tarball
* This is the internal Dockerfile so it cannot be used to publish the official image

A benefit would be that the generated `Dockerfile.local` could be more easily reviewed during a VOTE. However, I think we could add to the smoketester a test that builds the image, starts it, indexes some docs and stops? For this it is enough with the Dockerfile embedded in the binary distro.

Reply:
@HoustonPutman This ties in with how we publish our docker images for 9.x, and how we test and review them as part of the Release Process and VOTING. Technically we don't need to VOTE on the Dockerfile itself, only the solr tarball, so implicitly the VOTE also covers the Dockerfile.internal that is part of the release :) But to which extent do we want the release process to include automatic and maual tests of the Dockerfile and Docker image itself, and how do we do that?

Reply:
You don't need a separate folder for the local dockerfile (since as you mention, it's already in the tarball), but I think we should upload the proposed official dockerfile as a part of the release. That way we can build a test "official" image as a part of the smoke tester. Of course, this official dockerfile might change when we submit the PR to the official-images repo. But if we include it in the smoke tester, we have a much better idea that it is safe to release.

We can (and should) test both the `Dockerfile.local` as well as the `Dockerfile.official` as a part of the release IMO. It should be very straightforward and easy, and shouldn't take much time. I agree that we don't need to full suite of docker tests that we do on `gradle dockerTest`, we can just have a simple integration test (and do it for the images created from both dockerfiles).

The local dockerfile should be especially easy to test, because as you mention, it's included in the release TGZ

The official dockerfile should also be easy to test, since we host the release TGZ for others to download during the vote. We can just have the smoke tester build the docker image using a different release server (the one hosting the release candidate).